File: python-examples/cdp-connection/utils/types_and_schemas.py
from pydantic import BaseModel


class ConnectToCdpParams(BaseModel):
    """Parameters for connect-to-cdp API"""

    url: str


class BrowserInfo(BaseModel):
    """Browser information from CDP endpoint"""

    browser_version: str
    protocol_version: str
    user_agent: str
    web_socket_debugger_url: str  # WebSocket URL as string (ws:// or wss://)


class PageInfo(BaseModel):
    """Page information"""

    title: str
    url: str  # Page URL as string


class WebDriverInfo(BaseModel):
    """WebDriver capabilities and session info"""

    capabilities: dict
    session_id: str | None = None


class CDPConnectionResult(BaseModel):
    """Result returned from CDP connection API"""

    message: str
    cdp_url: str  # CDP URL can be HTTP or WS, use string for flexibility
    browser_info: BrowserInfo
    page_info: PageInfo
    web_driver_info: WebDriverInfo

File: typescript-examples/cdp-connection/utils/typesAndSchemas.ts
import { z } from "zod";

// Params schema for connect-to-cdp API
export const connectToCdpParamsSchema = z.object({
  url: z.string().url("Must be a valid URL").min(1, "URL is required"),
});

// Browser Info schema
export const browserInfoSchema = z.object({
  browserVersion: z.string(),
  protocolVersion: z.string(),
  userAgent: z.string(),
  webSocketDebuggerUrl: z.string().url(),
});

// Page Info schema
export const pageInfoSchema = z.object({
  title: z.string(),
  url: z.string().url(),
  viewport: z.object({
    width: z.number(),
    height: z.number(),
  }),
});

// WebDriver Info schema
export const webDriverInfoSchema = z.object({
  capabilities: z.record(z.any()),
  sessionId: z.string().optional(),
});

// CDP Connection Result schema
export const cdpConnectionResultSchema = z.object({
  message: z.string(),
  cdpUrl: z.string().url(),
  browserInfo: browserInfoSchema,
  pageInfo: pageInfoSchema,
  webDriverInfo: webDriverInfoSchema,
});

// Type exports inferred from schemas
export type ConnectToCdpParams = z.infer<typeof connectToCdpParamsSchema>;
export type BrowserInfo = z.infer<typeof browserInfoSchema>;
export type PageInfo = z.infer<typeof pageInfoSchema>;
export type WebDriverInfo = z.infer<typeof webDriverInfoSchema>;
export type CDPConnectionResult = z.infer<typeof cdpConnectionResultSchema>;
